Engine/platform/windows/gfx/ali3dd3d.cpp

```diff
@@ -1335,10 +1335,53 @@ void D3DGraphicsDriver::_render(bool clearDrawListAfterwards)
     _filter->SetSamplerStateForStandardSprite(direct3ddevice);
     D3DTEXTUREFILTERTYPE filterType;
     direct3ddevice->GetSamplerState(0, D3DSAMP_MAGFILTER, (DWORD*)&filterType);
-    if (direct3ddevice->StretchRect(pNativeSurface, NULL, pBackBuffer, &viewport_rect, filterType) != D3D_OK)
+    
+    const int pixelSize = viewport_rect.right / _srcRect.GetWidth();
+
+    if (pixelSize == 1) {
+        // since the pixel size is 1 the distortion will not happen
+        if (direct3ddevice->StretchRect(pNativeSurface, NULL, pBackBuffer, &viewport_rect, filterType) != D3D_OK)
+        {
+            throw Ali3DException("IDirect3DSurface9::StretchRect failed");
+        }
+    } 
+    else // TODO find a way to make this workaround unnecessary
     {
-      throw Ali3DException("IDirect3DSurface9::StretchRect failed");
+        // Bug: StretchRect makes the last right and bottom pixel one single pixel, always, no matter how big you stretch (reason uknown)
+        // HACK: we draw the surface in four parts to ensure the last row and column have the proper size
+        const int hackOffset = pixelSize - 1;
+        const int src_width = _srcRect.GetWidth();
+        const int src_height = _srcRect.GetHeight();
+
+        const RECT rectsHack[4][2] = {
+            {   // main area
```

Comment:
Is the main area still drawing a 1 pixel slice on the margins?

Reply:
Yes, that's why I'm accounting for one extra pixel